plugins/UsersManager/AuthTokenProvider.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\UsersManager;
+
+use Piwik\Common;
+use Piwik\Config;
+use Piwik\Date;
+use Piwik\Db;
+use Piwik\Plugins\UsersManager\AuthTokenNotifications\AuthTokenNotification;
+use Piwik\Plugins\UsersManager\AuthTokenNotifications\AuthTokenProviderInterface;
+use Piwik\Plugins\UsersManager\Model as UserModel;
+
+class AuthTokenProvider implements AuthTokenProviderInterface
+{
+    /** @var Model */
+    private $userModel;
+
+    /** @var string */
+    private $today;
+
+    public function __construct()
+    {
+        $this->userModel = new UserModel();
+        $this->today = Date::factory('today')->getDatetime();
+    }
+
+    private function getRotationPeriodThreshold(): string
+    {
+        $periodDays = Config::getInstance()->General['auth_token_rotation_notification_days'];
+        return Date::factory('today')->subDay($periodDays)->getDateTime();
+    }
+
+    public function setTokenNotified(string $tokenId): void
```

Comment:
As mentioned above, I would suggest to have the `setTokenNotified` part of the interface and always call it after a token was notified.

Reply:
I wanted to avoid the provider knowing about how to set tokens being notified, but now it is within the notification class itself and that doesn't need to know about that either, so I can move it around. At some point it would be good to perhaps have a separate model for tokens as currently they are just a side thing for users and don't have a solid structure of their own.